src/uu/cp/src/copydir.rs

```diff
@@ -0,0 +1,308 @@
+    if !source.is_dir() {
+        if preserve_hard_links {
+            let mut found_hard_link = false;
+            preserve_hardlinks(hard_links, &source, &local_to_target, &mut found_hard_link)?;
+            if found_hard_link {
+                return Ok(());
+            }
+            match copy_file(&source, &local_to_target, options, symlinked_files) {
+                Ok(_) => return Ok(()),
+                // silent the error with a symlink
+                // In case we do --archive, we might copy the symlink
+                // before the file itself
+                Err(_) if is_symlink(source) => return Ok(()),
+                Err(err) => return Err(err),
+            }
+        } else {
+            return copy_file(&source, &local_to_target, options, symlinked_files);
+        }
+    }
+
+    // In any other case, there is nothing to do, so we just return to
+    // continue the traversal.
+    Ok(())
+}
+
+/// Read the contents of the directory `root` and recursively copy the
+/// contents to `target`.
+///
+/// Any errors encountered copying files in the tree will be logged but
+/// will not cause a short-circuit.
+pub(crate) fn copy_directory(
+    root: &Path,
+    target: &TargetSlice,
+    options: &Options,
+    symlinked_files: &mut HashSet<FileInformation>,
+) -> CopyResult<()> {
+    if !options.recursive {
+        return Err(format!("omitting directory {}", root.quote()).into());
+    }
+
+    // if no-dereference is enabled and this is a symlink, copy it as a file
+    if !options.dereference && is_symlink(root) {
+        return copy_file(root, target, options, symlinked_files);
+    }
+
+    // check if root is a prefix of target
+    if path_has_prefix(target, root)? {
+        return Err(format!(
+            "cannot copy a directory, {}, into itself, {}",
+            root.quote(),
+            target.quote()
+        )
+        .into());
+    }
+
+    let mut hard_links: Vec<(String, u64)> = vec![];
+    let preserve_hard_links = options.preserve_hard_links();
+
+    // Collect some paths here that are invariant during the traversal
+    // of the given directory, like the current working directory and
+    // the target directory.
+    let context = match Context::new(root, target) {
+        Ok(c) => c,
+        Err(e) => return Err(format!("failed to get current directory {}", e).into()),
+    };
+
+    // Traverse the contents of the directory, copying each one.
+    for direntry_result in WalkDir::new(root)
+        .same_file_system(options.one_file_system)
+        .follow_links(options.dereference)
+    {
+        match direntry_result {
+            Ok(direntry) => {
+                let entry = Entry::new(&context, &direntry)?;
+                copy_direntry(
+                    entry,
+                    options,
+                    symlinked_files,
+                    preserve_hard_links,
+                    &mut hard_links,
+                )?;
```

Comment:
Is it okay to bubble the errors up here? Shouldn't they go to the `show_error!` on lin 303?

Reply:
I was matching the previous behavior. See the main loop in the old code, starting with:
```

    for path in WalkDir::new(root)
        .same_file_system(options.one_file_system)
        .follow_links(options.dereference)
    {
```
There's a bunch of `?` and `return Err(...)` that will terminate the loop immediately and exit the `copy_directory()` function with an error. I didn't check whether this is the right behavior when compared with GNU cp, but it seemed to be what the old code was doing. The `show_error!()` on line 303 corresponds to the `or_continue!()` macro in the old code.